Genomics
--------
- A 78-hr workshop on high-throughput sequencing analysis. We only had 12 hrs; if we had 2 wks we would have taught this: https://github.com/hbctraining/In-depth-NGS-Data-Analysis-Course
- `RNA-seqlopedia <https://rnaseq.uoregon.edu/>`_, an overview of RNA-seq from experimental design through analysis steps.
- Bioconductor's `RNA-seq workflow
  <https://www.bioconductor.org/help/workflows/rnaseqGene/>`_, walking you
  through a complete example analysis using DESeq2, including PCA plots and
  heatmaps.
- `"Learn Bioinformatics in 100 hours"
  <https://www.biostarhandbook.com/edu/course/1/>`_, an adaptation of the
  Applied Bioinformatics course from Penn State. **Costs $25** for the course
  and the Biostar Handbook. Well worth it if you're interested in learning
  bioinformatics further.
- `Ten simple rules for biologists learning to program
  <http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005871>`_,
  from PLOS Computational Biology
- `"Top N Reasons to do a PhD or post-doc in bioinformatics"
  <https://caseybergman.wordpress.com/2012/07/31/top-n-reasons-to-do-a-ph-d-or-post-doc-in-bioinformaticscomputational-biology/>`_,
  a blog post on why you would want to do this in the first place.
- `Applied Computational Genomics course
  <https://github.com/quinlan-lab/applied-computational-genomics>`_ from
  University of Utah, taught by Aaron Quinlan (author of BEDTools)
- The Harvard Chan Bioinformatics Core has several fantastic modules: https://github.com/hbctraining/main.
- `Introduction to Computational Biology
  <https://biodatascience.github.io/compbio>`_, a course taught by Mike Love
  (author of DESeq2). Lots of great links in his reading list, as well.
- More in-depth BEDTools tutorial: http://quinlanlab.org/tutorials/bedtools/bedtools.html
- BEDTools documentation: http://bedtools.readthedocs.io/en/latest/index.html
- The workflow we wrote and use for LCDB: https://github.com/lcdb/lcdb-wf. This
  runs RNA-seq and ChIP-seq analyses in an automated fashion on the biowulf
  cluster.

BASH
----
- Software Carpentry's `The Unix Shell <https://swcarpentry.github.io/shell-novice/>`_
- Codecademy's `Learn the Command Line <https://www.codecademy.com/learn/learn-the-command-line>`_
- The rather extensive `Bash guide <http://tldp.org/LDP/Bash-Beginners-Guide/html/>`_
- Environment variables and ``$PATH``: https://www.datacamp.com/community/blog/environment-variable-data-science

R
-
- `"swirl" <http://swirlstats.com/>`_, an interactive R tutorial 
- Guide to `customizing ggplot2 plots <http://r-statistics.co/Complete-Ggplot2-Tutorial-Part2-Customizing-Theme-With-R-Code.html#top>`_
-  `DataCamp <https://www.datacamp.com/>`_ has some R tutorials, and much more in-depth lessons if you pay
- The `R Inferno <http://www.burns-stat.com/pages/Tutor/R_inferno.pdf>`_, a tour throught the ugly parts of R told in a format similar to Dante's Inferno
- `Bioconductor for Genomic Data Science
  <https://kasperdanielhansen.github.io/genbioconductor/>`_, materials for
  a Coursera course. Great intro to the Bioconductor family of R packages.
- `DESeq2 tutorial <http://www.bioconductor.org/packages/3.7/bioc/vignettes/DESeq2/inst/doc/DESeq2.html>`_ from the authors of DESeq2.

Statistics
----------
- Textbook, "Broadening Your Statistical Horizons": https://bookdown.org/roback/bookdown-bysh/

UCSC Genome Browser
-------------------
- `UCSC Genome Browser training page <http://genome.ucsc.edu/training/index.html>`_
- `Open Helix training videos <http://www.openhelix.com/ucsc>`_
- UCSC `file formats page <http://genome.ucsc.edu/FAQ/FAQformat.html#format1>`_, with examples to try

